### PR TITLE
Optimize net.Read/WriteTable

### PR DESCRIPTION
**Pull request by meepdarknessmeep**
Last: https://github.com/garrynewman/garrysmod/pull/913

### CODE REVIEW

**Comment by Exho1** [comment]
This is good,the current ones are bloated as fuck

**Comment by DaaOtt** [comment]
+support

**Comment by meharryp** [comment]
Nice job, this really should make it in.

**Comment by Authorjames** [comment]
+Support, this should really make it in to the game.

**Comment by LennyPenny** [comment]
soo good

**Comment by Nyaaaa** [comment]
The most ridiculous lua code i've ever seen in my entire life!

**Comment by Exho1** [comment]
Have you done any benchmarks comparing this to the current system?

**Comment by wiox** [comment]
> Have you done any benchmarks comparing this to the current system?

There's one @ https://github.com/garrynewman/garrysmod/pull/913

**Comment by Exho1** [comment]
Thats what I meant, sweet

**Comment by Exho1** [comment]
This errors when you try to network a table with integer keys that are not entirely consecutive.

A table like this:
1	=	Player [1][Exho]
2	=	Player [1][Exho]
4	=	2
5	=	false

Throws this error:
[ERROR] addons/npc interaction/lua/thirdparty/net_meepdarkness.lua:714: bad argument #1 to 'WriteUInt' (number expected, got nil)
  1. WriteUInt - [C]:-1
   2. WriteTable - addons/npc interaction/lua/thirdparty/net_meepdarkness.lua:714
    3. sendResponse - addons/gphone/lua/gphone/sh_util.lua:247
     4. optionFunction1 - addons/gphone/lua/gphone/sh_util.lua:207
      5. DoClick - addons/gphone/lua/gphone/cl_animations.lua:722
       6. unknown - lua/vgui/dlabel.lua:218

**Comment by meepdarknessmeep** [comment]
I will update it here, I forgot to update it here after I changed it on gmodutil.

**Comment by robotboy655** [comment]
This is why you don't merge PRs instantly.

**Comment by meepdarknessmeep** [comment]
Nevermind, you are using an outdated version.

**Comment by Exho1** [comment]
I guess so, I just copied from this PR again (where I got it initially) and am not getting the error. I definitely copied it after your latest commit though so I'm not sure what went wrong. I'll keep my eyes out

**Comment by meepdarknessmeep** [comment]
Now that I look at this, it's terrible. Will be updating.

**Comment by meepdarknessmeep** [comment]
How does it look now?

File: garrysmod/lua/includes/extensions/net.lua [hunk 1]
```diff
@@ -71,111 +77,766 @@ end
...
+        local compressed = util.Compress( x )
+        
+        if ( compressed and compressed:len() < x:len() + 1 ) then
+        
+            net.WriteBool( true )
+            writing.uintv( compressed:len() )
+            net.WriteData( compressed, compressed:len() )
+            
+        else
+
```
Comment (Python1320) [inline garrysmod/lua/includes/extensions/net.lua]: Why is here no net.WriteBool( false )?

**Comment by Bo98** [comment]
What's with the change from tabs to 4 spaces?

**Comment by meepdarknessmeep** [comment]
is it a problem?

**Comment by Bo98** [comment]
It's inconsistent to the rest of the code base.

**Comment by Exho1** [comment]
So yes, thats a problem.

**Comment by meepdarknessmeep** [comment]
Is that better? I should've realized that before. Sorry.

**Comment by DaaOtt** [comment]
:+1:

**Comment by thegrb93** [comment]
And you called my code garbage. Shame on you.

**Comment by meepdarknessmeep** [comment]
First off, my code isn't garbage, second off your code is garbage.

Get off my pr, I don't even like you. You make terrible prs and all you do is try to act like a moderator and criticize others while you are at it.

**Comment by thegrb93** [comment]
Before you fixed it, it was, and you keep throwing that baseless insult at me, but you have never once pointed out where or why. Stop acting like a child.

**Comment by MrUnknownGamer** [comment]
Will you two shut the hell up. You two are arguing over fucking code. Not $1 Million, code. For gods sake, calm the hell down and just wait for someone to PR it.

**Comment by meepdarknessmeep** [comment]
Whats your problem? All I said was that he is way out of line. He is acting like a stubborn fuck, implying my code was horrible. All that was bad about it was formatting, then he comes in to tell me I should be shameful for calling his code shit and then making a request that has actual use.

So please stop budding into this conversation.


---- MrUnknownGamer wrote ----

**Comment by thegrb93** [comment]
Yes that's all you said is I'm way out of line. You are an angry hypocrite.

**Comment by MrUnknownGamer** [comment]
Just stop. Meep, you started it again by saying how your write by quoting my post. Just stop. 

I don't want to see another comment unless it is about the PR. If your pissed, use the thing called steam chat or PM on the forums.


Now all shut the hell up and no more comments unless it is about the PR

**Comment by MrUnknownGamer** [comment]
Right*

**Comment by meepdarknessmeep** [comment]
You dont fucking control me, who the fuck are you

---- MrUnknownGamer wrote ----

**Comment by meepdarknessmeep** [comment]
Im not angry at you nor am I a hypocrite. Literally no one would ever use your code. Its also terribly implemented.

---- thegrb93 wrote ----

**Comment by DaaOtt** [comment]
Well I mean starting a shitstorm works for getting a PR merged.

**Comment by meepdarknessmeep** [comment]
I wouldnt be talking to either the guy telling me what I can or cqnt do like he owns me or gdb if he wasnt acting like a butt earlier.

---- Ott wrote ----

**Comment by SwadicalRag** [comment]
Is there any reason this isn't merged yet?

This PR improves networking. There's no doubt about it.
The code looks good and is reasonably consistent with the rest of the repo.

Meep has put in a lot of thought and effort into this, and it at least deserves a response of some sort from the developers.
It's time to stop the net.WriteTable meme. A networking function shouldn't be as funny as Player:IsVehicle

**Pull request closed without merging**